From ddca4914d2c6f7bd6e9f552a751434b5c8323134 Mon Sep 17 00:00:00 2001
From: Avi Niddam <aniddam@habana.ai>
Date: Thu, 05 Jun 2025 16:35:50 +0300
Subject: [PATCH] [SW-228856] fit: use of temporary fct

Context:
We encountered several code duplications related to
fit_foreach_property.
So we created a new fit services infrastructure
to solve this issue in a clean way.
The main Lego pieces are the newly created files:
fit_services.c/h

Current commit:
In this current commit, we remove the static function
fit_foreach_property (which was duplicated code)
and use the function _fit_foreach_property instead.

General remark:
This infrastructure is intended to support ANY of the
fit related functions that we find today in fit_utils.c
or fit.c
Therefore, from now, please, anyone doing a change related
to the fit files should carefully double-check:
Any unified code can be (and should be) located in
fit_services.c/h
For this, please, follow the example of how it has been
done with fit_foreach_property.

Change-Id: Ib86a5c2af38a117f9e6cde2545366fb703b29839
Signed-off-by: Avi Niddam <aniddam@habana.ai>
---

diff --git a/lib/common/fit_utils.c b/lib/common/fit_utils.c
index 5dce23b..0cd864e 100644
--- a/lib/common/fit_utils.c
+++ b/lib/common/fit_utils.c
@@ -10,10 +10,13 @@
 #include <libfdt.h>
 
 #include <fit_utils.h>
+#include "fit_services.h"
 #include <emb_cmn_log.h>
 
 #ifdef __ZEPHYR__
 LOG_MODULE_REGISTER(fit_utils);
+#else
+#include <baremetal.h>
 #endif
 
 static bool fit_resp_node_end;
@@ -78,38 +81,12 @@
 	return FIT_FORMAT_NO_PROPERTY;
 }
 
-static int fit_foreach_property(const void *fit, const char *node_path, const char *match_prop)
-{
-	int ret, count = 0;
-	int noffset, property;
-	const struct fdt_property *prop;
-
-	noffset = fdt_path_offset(fit, node_path);
-	if (noffset < 0) {
-		EMB_LOG_RAW_ERR("Cannot find entry ");
-		EMB_LOG_RAW(node_path);
-		EMB_LOG_RAW("\n\r");
-		return noffset;
-	}
-
-	fdt_for_each_property_offset(property, fit, noffset) {
-		prop = fdt_get_property_by_offset(fit, property, NULL);
-		if (!prop)
-			continue;
-
-		ret = fit_match_prop_and_print(fit, prop, match_prop);
-		if (!ret)
-			count++;
-	}
-
-	return count;
-}
-
 void fit_print_versions(void *fit)
 {
 	int ret;
 
-	ret = fit_foreach_property(fit, "/", HL_FIT_VERSION);
+	ret = _fit_foreach_property(fit, "/", (_fit_property_match_cb)fit_match_prop_and_print,
+				    HL_FIT_VERSION);
 	if (ret <= 0)
 		EMB_LOG_WRN("[FIT] No version properties were found.");
 }
